[ToxOptimism] iP

ToxOptimism opened the pull request:
# Aurora 
>"When does the comet become a meteor?
> When does a candle become a blaze?
> When does a man become a monster?"
> \- "Just A Man" EPIC: The Musical

## Aurora is a task management ~~application~~ ***chatbot*** that is
- text-based
- user-friendly
- easy to set-up

## Set-up
1. download the [Aurora](https://github.com/ToxOptimism/ip/releases/tag/A-Jar) release file.
2. double-click it or cd to the directory Aurora is downloaded and run java -jar aurora-v0.1.jar in the directory
3. add your tasks.
4. let Aurora handle the rest! :purple_heart:

## Aurora's suite of features:
- [x] Add todos, events, deadlines
- [x] List tasks
- [x] Mark tasks as done
- [x] Delete tasks
- [ ] Find tasks by description
- [ ] Use a graphical user interace

## Here's the main method of Aurora:
```java
public static void main(String[] args) {

    try {
        storage.generateTaskListFile();
        loadTaskList();
    } catch (AuroraException e) {
        Ui.printMsg(e.getMessage());
        return;
    }

    Scanner sc = new Scanner(System.in);
    Command command = null;

    // Greet User
    Ui.printMsg(GREETING);

    // Chatbot
    while (command == null || !command.isExitCommand()) {
        String input = sc.nextLine().trim();
        try {
            command = Parser.of().parseCommand(input);
            command.execute(taskList, storage);
        } catch (AuroraException e) {
            Ui.printMsg(e.getMessage());
        }
    }

}
```

*You can take a look at Aurora's github [here](https://github.com/ToxOptimism/ip).*

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ToxOptimism committed:
Renamed Chatbot, Chatbot Greets and Exits

ToxOptimism committed:
Added bye command & echo user input.

ToxOptimism committed:
Added add tasks to tasklist & display tasklist, Removed echo

ToxOptimism committed:
Added Task class

ToxOptimism committed:
Added list command now using Tasks, added mark and unmark task.

ToxOptimism committed:
Added Deadline class

ToxOptimism committed:
Added ToDo class

ToxOptimism committed:
Added Event class

ToxOptimism committed:
Added todo, deadline, event commands, adjusted add task command to match.

ToxOptimism committed:
Added additional printed details when creating a task of a specific type.

ToxOptimism committed:
Added valid test cases till level 4

ToxOptimism committed:
Added exception handling, Added exception throwing for events, Updated test cases

ToxOptimism committed:
Added exception throwing for deadlines, Updated test cases

ToxOptimism committed:
Added exception throwing for todos, Updated test cases

ToxOptimism committed:
Added exceptions for mark and unmark, Updated test cases, Fixed an issue with index of list not printing correctly

ToxOptimism committed:
Added exception for unknown command, Updated test cases Removed unused function addTask and printException (printMsg was used instead)

ToxOptimism committed:
Added additional checks to prevent invalid arguments such as "/fromtest" rather than "/from", from working, Updated test cases

ToxOptimism committed:
Added delete, Added index not an int exceptions for mark, unmark and delete, Updated test cases

ToxOptimism committed:
Add generateTaskListFile function

There is no functionality to ensure the existence of the taskList file
or its parent directory.

This results in potential errors when attempting to read or write to the
taskList file if the required file or directory does not exist.

Let's,
* check for the existence of the task list file and its directory
* create the file and directory if they do not already exist

ToxOptimism committed:
Add loadTaskList function

There is no functionality to load tasks from taskList file into the
application.

This results in the application being unable to retrieve and initialize
tasks from previously saved data.

Let's,
* read taskList file
* parse each line to appropriate task type
* mark tasks as done if indicated

ToxOptimism committed:
Bug Fix: Adjust loadTaskList try-catch scope

Try-catch of loadTaskList function spanned the function, possibly
catching errors generated outside of the intended scope of reading file
lines.

Let's
* extract task type parsing from try-catch

ToxOptimism committed:
Add save to taskList file

There is no functionality to save task items into the data file.

This results in the application being unable to retrieve data between
runs.

The save functionality needs to be able to account for different task
operations such as creating new entry, removing entry and marking as
done.

Let's
* add an overwrite file function
* add an append file function
* add a method for tasks to parse into the appropriate file format

ToxOptimism committed:
Add usage of save task

The code did not utilize the overwriteTaskListFile and the
appendTaskListFile function.

This resulted in it not saving the task list file.

Let's
* Add usage of append to addTask function
* Add usage of overwrite to delete, mark and unmark functions

ToxOptimism committed:
Merge branch 'branch-Level-7'

ToxOptimism committed:
Convert deadline & events to use LocalDateTime

Deadlines and events use strings to represent datetime. This results in
difficulty handling and manipulating them.

Utilizing LocalDateTime would ensure ease in formatting.

Let's
* Update deadline and event class to use LocalDateTime
* Add checks when adding tasks to ensure format dd/mm/yyyy hhmm is used
* Update text output to use a format identical to Oct 2 2019 2:10pm
* Update file output to use dd/mm/yyyy hhmm for ease of parsing on load

ToxOptimism committed:
Update test cases

Newly added local date time did not match with test cases, resulted in
failed test cases.

Let's
* Update test cases to use localDateTime
* Include new cases to check for invalid date time inputs

ToxOptimism committed:
Merge branch 'branch-Level-8'

ToxOptimism committed:
TaskList class: extract tasklist behaviour

TaskList functions is handled by main file Aurora.java and coupled with
other behaviour such as saving and displaying messages.

Extracting the tasklist behaviour could improve code readability.

Let's
* Create a new class taskList
* Extract add, delete from taskList
* Extract mark, unmark tasks
* Extract convert tasklist to file format and display format

ToxOptimism committed:
Storage class: extract storage behaviour

Storage functions is handled by main file Aurora.java and coupled with
other behaviour such as displaying messages.

Extracting the storage behaviour could improve code readability.

Let's
* Create a new class Storage
* Extract generate data file
* Extract load data file
* Extract overwrite and append to data file

ToxOptimism committed:
Parser class: extract parser behaviour

Parser functions is handled by main file Aurora.java. These functions
are primarily utility functions that convert data into other data types.

Let's
* Create a new class Parser
* Extract parse lines from storage file
* Extract parse int
* Extract parse date time

ToxOptimism committed:
Ui class: extract Ui behaviour

Ui display functions is handled by Aurora.java.

Extracting these functions could improve code readability and reduce
coupling.

Let's
* Create a new class Ui
* Extract print msg
* Extract print taskList

ToxOptimism committed:
Add Command class: extract task creation commands

Creating new tasks requires parsing and executing the command, which is
done as functions in Aurora.java.

Extracting these functions could serve to abstract the commands of the
chatbot, thus improving code readability.

Let's
* Create a new Command class
* Create add tasks command classes
* Extract common calling of storage functionality (saving) as function
* Extract parsing of arguments specific to commands
* Extract execution of commands
* Extract common calling of UI upon adding task as a function

ToxOptimism committed:
Display Command classes: extract display commands

To standardize the level of abstraction of all commands, all commands
will be abstracted from Aurora.java, to improve code readability.

Let's
* Create a new ByeCommand class
* Create a new ListCommand class
* Extract execution of commands
* Extract exiting chatbot functionality as a command

ToxOptimism committed:
Edit List Command class: extract task modification commands

Editing tasks requires parsing of index and executing the command,
which is done as functions in Aurora.java.

To standardize the level of abstraction of all commands, all commands
will be abstracted from Aurora.java, to improve code readability.

Let's
* Create a new DeleteCommand class
* Create a MarkCommand & UnmarkCommand class
* Create add tasks command classes
* Extract parsing of arguments specific to commands
* Extract execution of commands

ToxOptimism committed:
Parser Class: Add parse commands

Identification of commands is handled by Aurora.java, when it should be
encapsulated under the Parser Class as part of its behaviour. This will
improve code readability as the functionality will be in a more
reasonable location.

Let's
* Extract parsing of commands

ToxOptimism committed:
Update Aurora.java to utilize new command classes

Removing outdated code that has been abstracted by new command classes.

Refactor Aurora.java to call parser & command class functionality. The
main functionality of Aurora.java should not concern itself with
identifying individual commands.

Let's
* Remove outdated code
* Use abstracted command identification

ToxOptimism committed:
Merge branch 'branch-A-MoreOOP'

ToxOptimism committed:
Add packages to organize classes

Classes are not categorized into their respective responsibilities.

Adding packages can help to ease the management and readability of
classes.

Let's
* Create packages - aurora, command, exception, task
* Create io package for interacting with input/ output/ storage
* Create util package for parser class
* Extract classes into respective packages

ToxOptimism committed:
Merge branch 'branch-A-Packages'

ToxOptimism committed:
Merge branch 'add-gradle-support'

ToxOptimism committed:
Update gradle file reference

Gradle application and shadowJar does not match with current file names

Let's
- Update gradle to reference the new file names

ToxOptimism committed:
Update Storage to singleton

Testing of adding commands would require a storage stub. Current static
class implementation of Storage would prevent the implementation of a
storage stub.

Singleton based storage would allow overriding of methods.

Commands can take in storage instance to differentiate Storage from its
stub equivalent.

Let's
- Update storage to a singleton
- Update commands to take in storage instance when executing
- Add storage stub class with minimal functionality

ToxOptimism committed:
Add JUnit add deadline test

Text UI test acts as system test, making identifying individual bugs
difficult.

Automated unit tests can help to identify issues in the code easily.

Let's
* Create a AddDeadlineCommandTest class to handle unit test

ToxOptimism committed:
Update Parser to singleton

Storage is a singleton, standardization of all classes in the same util
package would benefit code readability.

Let's
- Update parser to a singleton

ToxOptimism committed:
Add JUnit parser test

Text UI test acts as system test, making identifying individual bugs
difficult.

Automated unit tests can help to identify issues in the code easily.

Let's
* Create a ParserTest class to handle unit test

ToxOptimism committed:
Update gradle file for version

ToxOptimism committed:
Add automated checkstyle to gradle

ToxOptimism committed:
Add JavaDoc comments

Functions do not clearly indicate what they do.

JavaDoc comments would help in code readability.

Let's
* Add JavaDoc comments to all application classes

ToxOptimism committed:
Update code to match coding standards

Code does not align with automated checkstyle.

Let's
* Run the automated checkstyle
* Resolve any checkstyle issues

ToxOptimism committed:
Add find keyword in description

TaskList can only be listed in full.

Support for filtering task list by key words could improve the
usability of the application.

Let's
* Add a method to identify if a task has a keyword in its description.
* Add a method to list the tasks with the keyword.

ToxOptimism committed:
Add find command

The user has no means of specifying a keyword to filter by.

The addition of a find command, could improve usability of the
application.

Let's
* Add a find command.
* Allow the parser to recognize the command.

ToxOptimism committed:
Add find command tests

Find command does not have any corresponding unit or system tests.

Automated unit and system tests can help to identify issues in the code
easily.

Let's
* Add a parser test to verify parsing of find command.
* Add text ui test for valid and invalid use of the command.

ToxOptimism committed:
Update taskList java docs for clarity

ToxOptimism committed:
Merge branch 'branch-A-JavaDoc'

ToxOptimism committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/aurora/command/AddCommand.java
#	src/main/java/aurora/command/Command.java
#	src/main/java/aurora/io/Storage.java
#	src/main/java/aurora/task/Task.java
#	src/main/java/aurora/util/Parser.java

Resolution:
* Use master branch for most.
* For checkstyle issues, use coding standard branch.

ToxOptimism committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/aurora/task/Task.java
#	src/main/java/aurora/task/TaskList.java
#	src/main/java/aurora/util/Parser.java
#	src/test/java/aurora/util/ParserTest.java

Resolution:
* Fix JavaDocs comment for different functions being overlapped when
merging.
* Fix import order in accordance to checkstyle.

ToxOptimism committed:
Update gradle dependencies for javaFX

ToxOptimism committed:
Replace TUI with GUI

Text based UI can make the application feel unapproachable to new users
as they are unsure of what to do.

A familiar chatbot messaging style GUI would benefit user experience.

Let's
* Replace TUI with GUI based on Duke's JavaFX tutorial.
* Get input using the GUI.
* Display command responses using the GUI.
* Add a launcher class to launch the main application.
* Update the aurora application to load the UI scene.
* Update the aurora application to execute input on event.

ToxOptimism committed:
Add UiStub for test cases

The test cases for adding a new deadline fails as execution of command
is unable to access the Ui during test running.

This essential test cases if left unchanged might cause other issues
to be hidden.

A UiStub that can be referred to by the test cases would resolve the
issue.

Let's
* Add a UiStub with limited functionality for the purpose of testing

ToxOptimism committed:
Merge branch 'branch-Level-10'

iamanoob44 commented on src/main/java/aurora/command/AddCommand.java:
```diff
@@ -0,0 +1,29 @@
+     * @param taskList the taskList that stores the task.
+     * @param storage the storage to write to.
+     * @throws AuroraException if appending to storage fails.
+     */
```
I like how informative the JavaDoc comments are, nice!

iamanoob44 commented on src/main/java/aurora/command/AddCommand.java:
```diff
@@ -0,0 +1,29 @@
+     * @param storage the storage to write to.
+     * @throws AuroraException if appending to storage fails.
+     */
+    public void addToList(Task t, TaskList taskList, Storage storage) throws AuroraException {
```
Perhaps it would be good if you can rename your variable name 't' to something more descriptive such as 'taskToAdd' so that it follows the coding standard more closely? Overall, I still find your code easy to read and understand!

Since we usually allocate one single letter : `i`,`j` or `k` for for loops

ToxOptimism replied:
I will have to look into it, as due to the small scope of the function, a short name is appropriate for this situation. Additionally, this is standard with other methods such as TaskList's markTaskAsDone which references `Task t` as a scratch variable.  

Yes `i`, `j`, `k` are scratch variables usually reserved for for loops, likewise `m`, `n` are usually reserved for `int` input parameters and `c` and `d` references characters. As such, the code will likely be using `t` as a scratch variable for any variable of datatype Task.

iamanoob44 replied:
Ahh i see i see with regards to the scratch variable, but i wont deny the fact that the code still looks clean though. It's just i think perhaps is better to rename the variable 't' to 'task' for better clarity, so that it better aligns with the coding standard i guess?

iamanoob44 commented on src/main/java/aurora/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,97 @@
+        String beforeBy = info;
+
+        // Account for different combinations of /by usage
+        if (info.contains("/by ")) {
+            byDateStart = info.indexOf("/by ");
+            beforeBy = info.split("/by ")[0].trim();
+            description = beforeBy;
+        } else if (info.contains("/by\n")) {
+            byDateStart = info.indexOf("/by\n");
+            beforeBy = info.split("/by\n")[0].trim();
+        } else if (info.endsWith("/by")) {
+            byDateStart = info.length() - 3;
+            beforeBy = info.substring(0, info.length() - 3).trim();
+        }
+
+        // If there is no description provided
+        if (info.trim().isEmpty() || beforeBy.isEmpty()) {
+            throw new AuroraException("Missing argument: \"Description\".\n"
+                    + "Usage: \"deadline Description /by By\"");
+        }
+
+        // If there is no /by
+        if (byDateStart == -1) {
+            throw new AuroraException("Missing argument: \"/by By\"."
+                    + "\nUsage: \"deadline Description /by By\"");
+
+        // If there is no details after /to
+        } else if (byDateStart + 3 == info.length()) {
+            throw new AuroraException("Missing argument: \"By\" in \"/by By\".\n"
+                    + "Usage: \"deadline Description /by By\"");
+        }
```
Perhaps you can define certain repetitive hard-coded strings as constants to help increase your code maintainability and align with the coding standards more closely? For example, perhaps you can consider:

`private static final String BY_KEYWORD = "/by";`
`private static final String DEADLINE_DESCRIPTION = "Usage: \"deadline Description /by By\"";`

ToxOptimism replied:
Thanks for the suggestion! 😄 

Will likely implement this in future commits.

iamanoob44 commented on src/main/java/aurora/command/AddEventCommand.java:
```diff
@@ -0,0 +1,135 @@
+            throw new AuroraException("Missing argument: \"To\" in \"/to To\".\n"
+                    + "Usage: \"event Description /from From /to To\"");
+        }
+
```
I noticed a similar issue found in the `AddDeadlineCommand.java` file earlier. Perhaps you can try to define constants too for certain repetetive hard-coded strings?

iamanoob44 commented on src/main/java/aurora/exception/AuroraException.java:
```diff
@@ -0,0 +1,16 @@
+package aurora.exception;
+
+/**
+ * Represents an exception that occurs when there is an error in the aurora package.
+ */
+public class AuroraException extends Exception {
+
+    /**
+     * Constructs a new AuroraException with the specified message.
+     *
+     * @param message the error message detailing the cause of the exception.
+     */
+    public AuroraException(String message) {
+        super(message);
+    }
+}
```
LGTM, I like how you included javadoc comments for both the class and also for the constructor. I also like how clean your code is for this file!

iamanoob44 commented on src/main/java/aurora/task/TaskList.java:
```diff
@@ -0,0 +1,169 @@
+     * @return Task the task at specified index.
+     * @throws AuroraException if the index is out of bounds.
+     */
+    public Task getTask(int index) throws AuroraException {
+
+        validateIndex(index); // throws an exception
+
+        return taskList.get(index - 1);
+    }
```
Perhaps you can incorporate a 'try-catch' block to capture the `AuroraException`. Any reason why you did it this way instead of using a 'try-catch' ?

ToxOptimism replied:
Ah, the reasoning is that it is not the responsibility of the `getTask` function to handle the `AuroraException` when it is the caller that is supplying an invalid index. 

As such this exception is actually thrown up to any command that uses this function and in turn the command will typically throw it to the chatbot, since its also not the command's responsibility to handle invalid input. Thus the chatbot is the one to handle the try-catch capturing of the AuroraException.

Hope that clarifies things, and do you have any recommendation for comment improvements to bring additional clarity to this issue within the code?

iamanoob44 replied:
Ohhh i see, thanks for the clarification though and my bad on this then!

and with regards to the comment improvements, perhaps you can include your "validateIndex" part into the JavaDoc comments.

So perhaps something like this?

```Java
/**
 * Validates if the 1-based index is within the bounds of the list.
 * <p>
 *     The validateIndex method in {@code TaskList} also throws AuroraException when ...
 * </p>
 * @param index the 1-based index to be validated.
 * @throws AuroraException if the index is out of bounds.
 */
 ```

iamanoob44 reviewed (commented):
LGTM, just some nits to fix. Overall, great work with the code! it still closely aligns with the coding standards as I do not see any blatant violations of the coding standards for the course.

jeronkk commented on src/main/java/aurora/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,97 @@
+public class AddDeadlineCommand extends AddCommand {
+
+    // Deadline specific fields
+    private LocalDateTime bDate;
```
Maybe instead of bDate try byDate

bhavinaa replied:
Perhaps you could consider renaming the variable bDate. It is not very intuitive as to what that refers to. It would be hard for others to understand the code, the second time you see it

jeronkk commented on src/main/java/aurora/command/AddEventCommand.java:
```diff
@@ -0,0 +1,135 @@
+public class AddEventCommand extends AddCommand {
+
+    // Event specific fields
+    private LocalDateTime fDate;
```
Consider renaming it to fromDate and toDate

jeronkk commented on src/main/java/aurora/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,58 @@
+
+        super.execute(taskList, storage);
+
+        ToDo td = new ToDo(description);
```
maybe instead of td consider toDoTask

jeronkk commented on src/main/java/aurora/command/DeleteCommand.java:
```diff
@@ -0,0 +1,61 @@
+
+        super.execute(taskList, storage);
+
+        Task t = taskList.deleteFromList(index); // throws AuroraException if index is out of bounds
```
Can be named task instead of t

jeronkk commented on src/test/java/aurora/io/UiStub.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void printMsg(String msg) {
+        return;
+    }
```
Maybe write out the full word printMessage()

jeronkk reviewed (commented):
In all really impressed with your chatbot. Other than small naming conventions, really solid work!

bhavinaa commented on src/main/java/aurora/Aurora.java:
```diff
@@ -0,0 +1,91 @@
+package aurora;
```
This is my first PR, regarding the coding standard alone. So I would only be commenting on that. 

I like how you have organized your class, while grouping related packages, and at the same time you have restricted the accessibility, such that it is only accessible

bhavinaa commented on src/main/java/aurora/command/AddEventCommand.java:
```diff
@@ -0,0 +1,135 @@
+
+    // Event specific fields
+    private LocalDateTime fDate;
+    private LocalDateTime tDate;
```
I really like the javDoc comments, they are very descriptive and then explains the code well

bhavinaa commented on src/main/java/aurora/command/AddToDoCommand.java:
```diff
@@ -0,0 +1,58 @@
+     * @throws AuroraException the appropriate exception message if unable to parse arguments.
+     */
+    @Override
+    public void parseArgs(String[] argsList) throws AuroraException {
```
perhaps you could consider abstracting out the parsing to a different class, to reduce code repetition

bhavinaa commented on src/main/java/aurora/util/Parser.java:
```diff
@@ -0,0 +1,167 @@
+package aurora.util;
```
I like how you have specifically imported the functions you use from the common libraries instead of doing it with the *

bhavinaa reviewed (commented):
Overall, I feel you have really good code quality. You have consistent naming, and then your JavaDocs is really descriptive  making it easy for the new readers to look at it

ToxOptimism committed:
Update find command to display correctly

The find command displays the entire task list regardless of the filter.

Let's
*  Update find command's execute to display the filtered list

ToxOptimism committed:
Add assertion to various methods

The command, task, io classes assumes the arguments passed are not null.

The assertions will error if such assumptions are invalidated.

This guards against violating such assumptions in future iterations of
development.

Let's
* Add null assertion checks to the command classes
* Add null assertion checks to the task classes
* Add null assertion checks to the io classes

ToxOptimism committed:
Update command execute with assertions

The command class uses an exception to prevent the usage of commands
without parsing the necessary arguments. The command class uses an
exception to re-execution of commands.

The cases represent issues with ordering of function calls, a programmer
error.

This does not follow the convention of using exceptions for user input
errors and assertions for programmer errors.

Let's
* Replace the arguments not parsed exception with an assertion
* Replace the command already executed exception with an assertion

ToxOptimism committed:
Update variable names for clarity

Peer review has indicated that the usage of scratch variable naming
affected code readability.

Appropriate variable names could help improve the understanding of
developers involved in future iterations.

Let's
* Update scratch variables to utilize a more comprehensive name
* Update JavaDocs to match the parameter and return variable names

ToxOptimism committed:
Replace magic literals with constant

Commands, Tasks and Io's various exception messages, usage messages and
argument identifiers currently use various magic literals, leading to
duplicated values that are meant to reference the same variable.

Using named constants help improve the readability and maintainability
of the code.

Let's
* Replace exception messages literals with constants
* Use a command usage constant for all exceptions in a command
* Use a command keyword constant for the parser to identify commands
* Use a task keyword constant for the parser to identify tasks

ToxOptimism committed:
Add comments to justify code duplication

The peer review highlighted enquiries regarding the duplication of
select code in parsing of arguments.

However the code is designed such that it is easily extensible. As such,
the parsing of the arguments are coupled with the command such that
modifications to the command only affect that specific class.

Let's
* Add a comment to justify why the code is structured in such a manner

ToxOptimism committed:
Add not null assertion to more methods

The parser and aurora classes assumes the arguments passed are not null.

The assertions will error if such assumptions are invalidated.

This guards against violating such assumptions in future iterations of
development.

Let's
* Add null assertion checks to the aurora classes
* Add null assertion checks to the parser classes
* Remove test cases that checked if an assumption, covered by an
  assertion, was violated

ToxOptimism committed:
Merge pull request #1 from ToxOptimism/branch-A-Assertions

Merge branch-A-Assertions into master branch

ToxOptimism committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/aurora/command/AddCommand.java
#	src/main/java/aurora/command/AddEventCommand.java
#	src/main/java/aurora/command/AddToDoCommand.java
#	src/main/java/aurora/command/Command.java
#	src/main/java/aurora/command/DeleteCommand.java
#	src/main/java/aurora/command/FindCommand.java
#	src/main/java/aurora/command/MarkCommand.java
#	src/main/java/aurora/command/UnmarkCommand.java
#	src/main/java/aurora/task/TaskList.java

Resolution:
* Take both for most
* For issues with variable name, take code quality branch's
* Remove taskList's tasklist and task not null assertion, shift to when
  adding tasks

ToxOptimism committed:
Merge pull request #2 from ToxOptimism/branch-A-CodeQuality

Merge branch-A-CodeQuality into master branch

ToxOptimism committed:
Add DoWithinPeriod tasks

The set of deadlines, events and todos could be seen as limiting. Users
may find more options of task categories useful. A possible variation of
tasks would be DoWithinPeriod tasks, which indicates tasks that should
be done within a certain period.

Functionally, a DoWithinPeriod task would be identical to an Event task,
as the validation of whether a task is marked as done within a certain
time frame is out of scope. The distinguishing factor would be task
keyword that is displayed.

Notably, the argument format will be identical to an Event task as well.
However the parsing of arguments are to remain separate to avoid
coupling commands together for future changes.

Let's
* Add a DoWithinPeriod task
* Add a AddDoWithinPeriod command
* Add the command to Parser
* Update the version of the application

ToxOptimism committed:
Merge branch 'branch-BCD-Extension'

ToxOptimism committed:
Reduce length of parse arguments methods

The issue tracker has flagged parsing of arguments as too lengthy.

Finding of "/{argIdentifier}" in the input string can be extracted
as a method to reduce code duplication and code length.

Let's
* Add a protected method for finding an argument's start index by the
  argument identifier
* Add a protected method for checking if there is any other text before
  the argument identifier

ToxOptimism committed:
Update JavaDocs

Issue tracker indicated areas where the non-plural form was used for
method header comments' overview statement.

Let's
* Update all the JavaDoc's overview statements

ToxOptimism committed:
Fix exception check to account for whitespace

Event and DoWithinPeriod did not throw the correct exception in the case
of "/from    /to" and "/start    /end" due to the whitespaces.

Let's
* Account for whitespace by trimming the substring between the two
  identifiers

ToxOptimism committed:
Add Ui.png

ToxOptimism committed:
Update README.md

ToxOptimism committed:
Fix a bug with add event command

ToxOptimism committed:
Fix a bug with add do within period command

ToxOptimism committed:
Fix a bug with parser

ToxOptimism committed:
Update README.md
